Add recently accepted blocks and txn to AttemptToEvictConnection.

gmaxwell opened the pull request:
This protect any not-already-protected peers who were the most recent
to relay transactions and blocks to us.

This also takes increases the eviction agressiveness by making it
willing to disconnect a netgroup with only one member.

gmaxwell committed:
Add recently accepted blocks and txn to AttemptToEvictConnection.

This protects any not-already-protected peers who were the most
 recent four to relay transactions and most recent four to send
 blocks to us.

wallclockbuilder commented on src/net.cpp:
```diff
@@ -928,6 +950,20 @@ static bool AttemptToEvictConnection(bool fPreferNewConnection) {
+    if (vEvictionCandidates.empty()) return false;
+
+    // Protect 4 nodes that most recently sent us blocks.
+    // An attacker cannot manipulate this metric without performing useful work.
```
Ack

rebroad replied:
Actually, I think an attacker can manipulate this metric without performing useful work. It's quite trivial to send a block inv as soon as a node sees the block inv, and this can beat most other nodes before they're able to send the actual block. Meanwhile the average node will try to download the block from the node which sent the block inv first, rather than the fastest to send the block. This could mean that BOTH the block propagation is slowed down AND that the nodes responsible for this happening continue to be protected due to this code. A better solution would be to regularly test which of the connected nodes actually deliver the block quickest - perhaps by requesting blocks during quiet periods (when a new block hasn't just been released) and see which deliver it fastest. Also, it might be that the node which broadcasts the block inv first becomes one of the slowest nodes to relay the actual block due to the fact that so many nodes are requesting the block from it. Perhaps the 2nd or 3rd node that sent the block inv would be able to deliver the block quicker. Perhaps, depending on the bandwidth of the sending and the receiving node, it would be more efficient and safer to request the newly announced block from 2 or more nodes concurrently, and reward the node (in this code) that delivered it soonest. The bandwidth of nodes (ours and theirs) could be tested and this information used to determine the best way to request the blocks.

sipa commented on src/net.cpp:
```diff
@@ -852,6 +857,21 @@ static bool ReverseCompareNodeTimeConnected(const NodeEvictionCandidate &a, cons
 
+static bool CompareNodeBlockTime(const NodeEvictionCandidate &a, const NodeEvictionCandidate &b)
+{
+    if (a.nLastBlockTime != b.nLastBlockTime) return a.nLastBlockTime < b.nLastBlockTime;
```
Isn't this backward? This sorts by increasing last block time, and erases the last items.

gmaxwell replied:
Erasing protects them from being evicted, thus keeping the most recent sources.

sipa replied:
Ah!

CodeShark replied:
Can be attacked by sending the same block repeatedly.

gmaxwell replied:
Fixed. (@sipa is this what you suggested?)

sipa replied:
Indeed!

arowser commented:
Can one of the admins verify this patch?

sipa commented:
utACK e6d8aaed0e4cc2a82715969371a0c026a3fc47fd

sipa commented:
Needs rebase.

gmaxwell commented:
Rebased.

btcdrak commented:
needs rebase.

gmaxwell commented:
Rebased.

pstratem commented:
@gmaxwell Can you split this into two commits?

gmaxwell commented:
No problem.

pstratem commented on src/net.cpp:
```diff
@@ -854,7 +859,22 @@ static bool ReverseCompareNodeTimeConnected(const NodeEvictionCandidate &a, cons
+
+static bool CompareNodeBlockTime(const NodeEvictionCandidate &a, const NodeEvictionCandidate &b)
+{
+    if (a.nLastBlockTime != b.nLastBlockTime) return a.nLastBlockTime < b.nLastBlockTime;
```
Please add a comment here explaining why there is a fail through.

gmaxwell committed:
Allow disconnecting a netgroup with only one member in eviction.

With the latest additions there are enough protective measures that
 we can take the training wheels off.

pstratem commented:
ACK 3b3f45d532bd8f4c917ff78340b63e355da0cf13

instagibbs commented on src/net.cpp:
```diff
@@ -854,7 +859,24 @@ static bool ReverseCompareNodeTimeConnected(const NodeEvictionCandidate &a, cons
+    // There is a fall-through here because it is common for a node to have more than a few peers that have not yet relayed txn.
+    if (a.nLastTXTime != b.nLastTXTime) return a.nLastTXTime < b.nLastTXTime;
+    if (a.fRelayTxes != b.fRelayTxes) return b.fRelayTxes;
+    if (a.fBloomFilter != b.fBloomFilter) return a.fBloomFilter;
```
Is this meant to be activated later? I don't see it being set anywhere. Also I assume the field means "peer requests bloom filters"?

instagibbs commented:
utACK https://github.com/bitcoin/bitcoin/pull/8084/commits/3b3f45d532bd8f4c917ff78340b63e355da0cf13

gmaxwell commented:
Anything more needed here?

sdaftuar commented:
utACK

sipa commented:
utACK 6ee7f05622c32431a9815a96b31a6a65a821fdcc

pstratem commented:
ACK 6ee7f05622c32431a9815a96b31a6a65a821fdcc
